compiler/src/iree/compiler/InputConversion/TOSA/TosaToLinalgExt.cpp

```diff
@@ -0,0 +1,187 @@
+
+static SmallVector<utils::IteratorType> getNParallelLoopsAttrs(
+    unsigned nParallelLoops) {
+  return SmallVector<utils::IteratorType>(nParallelLoops,
+                                          utils::IteratorType::parallel);
+}
+
+// Converts tosa.scatter to the iree_linalg_ext.scatter operation. As the
+// LinalgExt version is not batched therefore we materialize the batch index
+// for each update.
+class ScatterConversion : public OpRewritePattern<tosa::ScatterOp> {
+ public:
+  using OpRewritePattern<tosa::ScatterOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::ScatterOp op,
+                                PatternRewriter &rewriter) const final {
+    auto values = op.getValuesIn();
+    auto indices = op.getIndices();
+    auto updates = op.getInput();
+    auto valuesTy = values.getType().cast<ShapedType>();
+    auto indicesTy = indices.getType().cast<ShapedType>();
+    auto updatesTy = updates.getType().cast<ShapedType>();
+    ImplicitLocOpBuilder builder(op.getLoc(), rewriter);
+
+    if (!valuesTy.hasRank() || !indicesTy.getRank() || !updatesTy.getRank())
+      return rewriter.notifyMatchFailure(op,
+                                         "tosa.gather has unknown input rank");
+
+    // Materialize the unary index dimension.
+    llvm::SmallVector<int64_t> expandIndShape{indicesTy.getDimSize(0),
+                                              indicesTy.getDimSize(1), 1};
+    SmallVector<ReassociationExprs, 4> expandIndMap;
+    expandIndMap.push_back({
+        builder.getAffineDimExpr(0),
+    });
+    expandIndMap.push_back({
+        builder.getAffineDimExpr(1),
+        builder.getAffineDimExpr(2),
+    });
+
+    indices = builder.create<tensor::ExpandShapeOp>(
+        indicesTy.clone(expandIndShape), indices, expandIndMap);
+    indicesTy = indices.getType().cast<ShapedType>();
+
+    // Materialize the batch indice as LinalgExt scatter is not batched.
+    {
+      llvm::SmallVector<Value> dynDims;
+      for (int i = 0; i < indicesTy.getRank(); i++)
+        if (indicesTy.isDynamicDim(i))
+          dynDims.push_back(builder.create<tensor::DimOp>(indices, i));
+
+      auto empty =
+          builder
+              .create<tensor::EmptyOp>(indicesTy.getShape(),
+                                       indicesTy.getElementType(), dynDims)
+              .getResult();
+
+      Value batchIdx = nullptr;
+
+      if (indicesTy.getDimSize(0) == 1) {
+        Value zero = builder.create<arith::ConstantOp>(
+            rewriter.getZeroAttr(indicesTy.getElementType()));
+        batchIdx = builder.create<linalg::FillOp>(zero, empty).getResult(0);
+      } else {
+        SmallVector<AffineMap, 3> indexingMaps(
+            2, builder.getMultiDimIdentityMap(indicesTy.getRank()));
+        batchIdx = builder
+                       .create<linalg::GenericOp>(
+                           indicesTy, indices, empty, indexingMaps,
+                           getNParallelLoopsAttrs(indicesTy.getRank()),
```

Comment:
Is this only usage of this function here? (its a one line function that is not too difficult to interpret as N parallel loops ... but I'm not going to argue strongly if you feel this is more readable).

Reply:
Yup, it was the only common. Its a pretty common helper but I removed for now.